security-csrf/src/main/java/io/micronaut/security/csrf/filter/CsrfFilter.java

```diff
@@ -0,0 +1,192 @@
+    private final CsrfTokenValidator<HttpRequest<?>> csrfTokenValidator;
+    private final ExceptionHandler<AuthorizationException, MutableHttpResponse<?>> exceptionHandler;
+    private final CsrfFilterConfiguration csrfFilterConfiguration;
+
+    CsrfFilter(CsrfFilterConfiguration csrfFilterConfiguration,
+               List<ReactiveCsrfTokenResolver<HttpRequest<?>>> reactiveCsrfTokenResolvers,
+               List<CsrfTokenResolver<HttpRequest<?>>> csrfTokenResolvers,
+               CsrfTokenValidator<HttpRequest<?>> csrfTokenValidator,
+               ExceptionHandler<AuthorizationException, MutableHttpResponse<?>> exceptionHandler) {
+        this.csrfTokenResolvers = csrfTokenResolvers;
+        this.reactiveCsrfTokenResolvers = reactiveCsrfTokenResolvers.isEmpty()
+                ? reactiveCsrfTokenResolvers
+                : ReactiveCsrfTokenResolver.of(csrfTokenResolvers, reactiveCsrfTokenResolvers);
+        this.csrfTokenValidator = csrfTokenValidator;
+        this.exceptionHandler = exceptionHandler;
+        this.csrfFilterConfiguration = csrfFilterConfiguration;
+    }
+
+    @Override
+    public Publisher<MutableHttpResponse<?>> doFilter(HttpRequest<?> request, ServerFilterChain chain) {
+        Supplier<Publisher<MutableHttpResponse<?>>> proceedRequest = () -> chain.proceed(request);
+        Supplier<Publisher<MutableHttpResponse<?>>> denyRequest = () -> Mono.just(unauthorized(request));
+        if (!shouldTheFilterProcessTheRequestAccordingToTheHttpMethod(request)) {
+            return proceedRequest.get();
+        }
+        if (!shouldTheFilterProcessTheRequestAccordingToTheContentType(request)) {
+            return proceedRequest.get();
+        }
+        return reactiveCsrfTokenResolvers.isEmpty()
+                ? imperativeFilter(request, proceedRequest, denyRequest)
+                : reactiveFilter(request, proceedRequest, denyRequest);
+    }
+
+    private Publisher<MutableHttpResponse<?>> reactiveFilter(HttpRequest<?> request,
+                                                               Supplier<Publisher<MutableHttpResponse<?>>> proceedRequest,
+                                                               Supplier<Publisher<MutableHttpResponse<?>>> denyRequest) {
+        return Flux.fromIterable(this.reactiveCsrfTokenResolvers)
+                .concatMap(resolver -> Mono.from(resolver.resolveToken(request))
+                        .filter(csrfToken -> {
+                            LOG.debug("CSRF Token resolved");
+                            if (csrfTokenValidator.validateCsrfToken(request, csrfToken)) {
+                                return true;
+                            } else {
+                                LOG.debug("CSRF Token validation failed");
+                                return false;
+                            }
+                        }))
+                .next()
+                .flatMap(validToken -> Mono.from(proceedRequest.get()))
```

Comment:
If I replace it to: 

```java
 .flatMap(validToken -> proceedRequest.get())
```

it complains with: 

```
no instance(s) of type variable(s) R exist so that Publisher<Optional<MutableHttpResponse<?>>> conforms to Mono<? extends R>
```

Reply:
you may have to give explicit type parameters, something like `.<HttpResponse<?>>flatMap(...)`

Reply:
resolved via https://github.com/micronaut-projects/micronaut-security/pull/1826/commits/4506c4748a809eba61ecb2ab9593f2fe702773dd